[ENH] HyperSpectra: display visible image

pisarik opened the pull request:
Hi!

Sorry, I didn't start discussion in issues, since we discussed it at meetings.

Here is the following changes:

- Save all visible (microscopic) images (bytes, title, position and pixel size) from opus file into `data.attributes: dict[List[dict]]`;
- Bytes unpacked into `np.ndarray` with pillow package just before displaying it in `imageplot` (to save memory);
- Visible image can be selected from combo box;
- User can change opacity of visible image;
- User can select composition mode (Normal, Overlay, Multiply, Difference);

I don't have a good lightweight file to put in tests for Opus. The one I have weights 150 MBs

Closes #118

markotoplak commented on orangecontrib/spectroscopy/data.py:
```diff
@@ -700,6 +700,17 @@ def read(self):
+                'pixel_size_x': img['PixelSizeX'],
+                'pixel_size_y': img['PixelSizeY'],
+            })
+
```
@stuart-cls, is this safe? Do images always contain these attributes or should we wrap in try ... except?

stuart-cls replied:
We should wrap. In opusFC I just directly copy the attributes from the file, so if Bruker has used different labels in the past/future, the name will change. Unfortunately, the visible images part is not documented so I had to do some guessing there.

markotoplak commented:
Thanks! Seem like a quality code, I will check more in depth if the settings (saving, loading, changing data sets) are handled as in other widgets.

Please rebase to recent master. I have just updated travis to use correct versions of scikit-learn so that the tests will run.

Regarding huge data files and testing. Well, this is only a problem when testing file readers. When testing HyperSpectra, you can always add an image manually by setting appropriate .attributes, and this is for testing, I think, even better. So there is no need to read an opus file when testing HyperSpectra.

borondics commented:
Hi @pisarik, thanks for this contribution it will be very useful. Could you also upload somewhere (Dropbox or similar) a few FPA images that we could use for human testing. We will also get data from other instruments and that should make sure that everything works fine. Also, if you could add a very small FPA file to do testing, it could be useful... or @markotoplak, should that be for opusFC?

markotoplak commented on orangecontrib/spectroscopy/tests/test_readers.py:
```diff
@@ -70,6 +72,37 @@ def test_roundtrip(self):
             np.testing.assert_equal(d1.X, d2.X)
 
 
+class TestOpusReader(unittest.TestCase):
```
Watch out with OPUS files, always add  `@unittest.skipIf(opusFC is None, "opusFC module not installed")` and also test without opusFC installed, because OpusFC is an optional library.

That is another reason to remove opus files from Hyper tests.

pisarik replied:
Agree that it is unnecessary to use opus files to test HyperSpectra. Since I store image bytes (jpeg or bmp) in attributes it is a little bit difficult to come up with dummy image bytes.

I can mock `Image.open` everywhere as it done in `test_select_another_visible_image`. Do you think it is the way to do it?

markotoplak replied:
Well, I'd prefer not to mock it Image.open, I would rather mock the data. You could perhaps add a small image to the repo and then read that one as bytes. Or, if the image is really small (few pixels would be enough anyway), you could base64 encode it and just leave it in the test .py file.

pisarik commented:
Hi @borondics! I've just sent to your email a link with the FPA file.

I asked to make a special measurement to obtain lightweight dummy Opus file for tests. But at this time I don't have it.

I also have a code to extract visible images from Agilent files. But I need to have more test files, since I hardcoded few parameters that I couldn't find in Agilent file. It perfectly works for the files that I have, but I would like to check it with Agilent files from other sources. Do you have Agilent FPA Image files?  It will greatly help me to extract visible image more reliably.

borondics commented:
Not sure if this is the reason, but HyperSpectra is broken for me using https://github.com/pisarik/orange-spectroscopy/tree/owhyper-show-visible-image and the FPA file @pisarik sent.

Here is the error message:
```
---------------------------- ValueError Exception -----------------------------
Traceback (most recent call last):
  File "/Users/borondics/Documents/code/Python/orange-canvas-core/orangecanvas/scheme/widgetmanager.py", line 378, in __process_init_queue
    self.ensure_created(node)
  File "/Users/borondics/Documents/code/Python/orange-canvas-core/orangecanvas/scheme/widgetmanager.py", line 326, in ensure_created
    self.__add_widget_for_node(node)
  File "/Users/borondics/Documents/code/Python/orange-canvas-core/orangecanvas/scheme/widgetmanager.py", line 242, in __add_widget_for_node
    w = self.create_widget_for_node(node)
  File "/Users/borondics/Documents/code/Python/orange-widget-base/orangewidget/workflow/widgetsscheme.py", line 296, in create_widget_for_node
    widget = self.create_widget_instance(node)
  File "/Users/borondics/Documents/code/Python/orange-widget-base/orangewidget/workflow/widgetsscheme.py", line 409, in create_widget_instance
    widget.__init__()
  File "/Users/borondics/Documents/code/Python/orange-spectroscopy/orangecontrib/spectroscopy/widgets/owhyper.py", line 919, in __init__
    self.setup_visible_image_controls()
  File "/Users/borondics/Documents/code/Python/orange-spectroscopy/orangecontrib/spectroscopy/widgets/owhyper.py", line 962, in setup_visible_image_controls
    callback=self.update_visible_image)
  File "/Users/borondics/Documents/code/Python/orange3/Orange/widgets/gui.py", line 476, in comboBox
    return gui_comboBox(*args, **kwargs)
  File "/Users/borondics/Documents/code/Python/orange-widget-base/orangewidget/gui.py", line 1498, in comboBox
    callfront.action(cindex)
  File "/Users/borondics/Documents/code/Python/orange-widget-base/orangewidget/gui.py", line 1938, in action
    raise ValueError("Combo box does not contain item " + repr(value))
ValueError: Combo box does not contain item -1

```

pisarik commented:
- Rebased onto master
- TestOpusReaders: added `skipIf`
- TestVisibleImage: replaced `opus/one_visible_image.0` with mocked images (base64 strings in the file)

pisarik commented:
> Not sure if this is the reason, but HyperSpectra is broken for me using https://github.com/pisarik/orange-spectroscopy/tree/owhyper-show-visible-image and the FPA file @pisarik sent.

@borondics Sorry, I used old version of Orange to develop. This error doesn't happen in Orange 3.24.0, but it indeed appeared when I pulled latest orange3 master branch as well. I think, you can checkout to the 3.24.0 to test manually.

pisarik commented:
Either this is a bug in orange-widgets or I don't know how to correctly initialize empty combo box.

I've explored the problem: 
**My code commit b8b3fac**
I initialize `vis_img_combo` with empty `PyListModel` (this will be filled with image names when data loaded):
```python
        self.vis_img_name_model = PyListModel()
        self.vis_img_combo = gui.comboBox(
            self.visbox, self, 'cur_visible_image_idx',
            model=self.vis_img_name_model)
```
Since Qt by default returns index `-1` when combo is empty, I initialize setting with this value:
```python
    cur_visible_image_idx = Setting(-1)`
```

**Orange codebase**
The error happens here, likely because it doesn't check if the model is empty before try to activate:

https://github.com/biolab/orange-widget-base/blob/35df1f0814b85de565437b359e33d30244698b40/orangewidget/gui.py#L1492-L1498

stuart-cls commented:
Thanks for this!

About the Agilent files, I have a (very) old prototype of this overlay for Agilent data. When this PR is merged, I'd be happy to add the support for Agilent files based on this PR.

pisarik commented:
@stuart-cls do you mean to remove from this PR `TestOpusReader` test suite, but leave changes to OpusReader?

stuart-cls commented:
That is what I meant, but I'd rather have the tests than not. Let's just make sure we don't conflate the hyperspectra tests with OpusReader (I think @markotoplak already covered this point well).

I can reproduce the combobox error with Orange-master but it's something to do with ContextSettings I think so I'll leave that for @markotoplak :)

markotoplak commented:
@pisarik, I have rebased this to master and changed the combo box to also support Orange 3.25+ (well, I dropped support for Orange 3.24). Please check if the widget still works properly. I might have missed something...

If you plan to develop this further, please continue on this branch.

The tests that only worked with big images were made optional, so not they pass.

Before merging we still need to:
- [ ] Wrap image reading code for Opus files in try-except, because those parameters might not be there in all files
- [ ] Fix BW image problems I already told you about.

Do you perhaps have time to do these shortly? If not, no problem, just tell me, I'll try to do them today/tomorrow.

You wrote beautiful code with tests. Thank you for your hard work.

pisarik commented:
Fixed two mentioned above issues:

- [x] Try-catch around image reading code
- [x] Cast all images to RGBA format

I have few questions:
1. Should I make interactive rebase to embed those commits in history and then force-push?
2. There is still a little bug: 
2.1 Connect test image data to HyperSpectra
2.2 Create another test image data and connect it to HyperSpectra
2.3 Then Show visible image is unchecked, but controls are enabled
3. Should I do something about PyLint problems?

I will try to fix bug with reconnection it.

Thanks a lot, I learned from your review as well!:)

markotoplak commented:
Thank you!  No need to do an interactive rebase in this case.

I can wrap this one now. I will remove two lint errors and will interactively rebase then in (I see that I added at least two of them yesterday).

And I already see a solution to your bug.
